protocols/relay/src/v2/protocol/outbound_hop.rs

```diff
@@ -0,0 +1,221 @@
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::v2::message_proto::{hop_message, HopMessage, Peer, Status};
+use crate::v2::protocol::{HOP_PROTOCOL_NAME, MAX_MESSAGE_SIZE};
+use asynchronous_codec::{Framed, FramedParts};
+use bytes::Bytes;
+use futures::{future::BoxFuture, prelude::*};
+use futures_timer::Delay;
+use libp2p_core::{upgrade, Multiaddr, PeerId};
+use libp2p_swarm::NegotiatedSubstream;
+use prost::Message;
+use std::convert::TryFrom;
+use std::io::Cursor;
+use std::iter;
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+use thiserror::Error;
+use unsigned_varint::codec::UviBytes;
+
+pub enum Upgrade {
+    Reserve,
+    Connect { dst_peer_id: PeerId },
+}
+
+impl upgrade::UpgradeInfo for Upgrade {
+    type Info = &'static [u8];
+    type InfoIter = iter::Once<Self::Info>;
+
+    fn protocol_info(&self) -> Self::InfoIter {
+        iter::once(HOP_PROTOCOL_NAME)
+    }
+}
+
+impl upgrade::OutboundUpgrade<NegotiatedSubstream> for Upgrade {
+    type Output = Output;
+    type Error = UpgradeError;
+    type Future = BoxFuture<'static, Result<Self::Output, Self::Error>>;
+
+    fn upgrade_outbound(self, substream: NegotiatedSubstream, _: Self::Info) -> Self::Future {
+        let msg = match self {
+            Upgrade::Reserve => HopMessage {
+                r#type: hop_message::Type::Reserve.into(),
+                peer: None,
+                reservation: None,
+                limit: None,
+                status: None,
+            },
+            Upgrade::Connect { dst_peer_id } => HopMessage {
+                r#type: hop_message::Type::Connect.into(),
+                peer: Some(Peer {
+                    id: dst_peer_id.to_bytes(),
+                    addrs: vec![],
+                }),
+                reservation: None,
+                limit: None,
+                status: None,
+            },
+        };
+
+        let mut encoded_msg = Vec::new();
+        msg.encode(&mut encoded_msg)
+            .expect("Vec to have sufficient capacity.");
```

Comment:
```suggestion
        let mut encoded_msg = Vec::with_capacity(msg.len());
        msg.encode(&mut encoded_msg)
            .expect("Vec to have sufficient capacity.");
```

We know how long the message is, best to do it in one allocation.

Reply:
Done across all of `protocol/` with be9b94ec70784492c6ca23ba7300e5dd0066462d.